File: apps/docker-server/cmd/renewlet/subscription_public_visibility_test.go
package main

import (
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"reflect"
	"strings"
	"testing"
	"time"

	"github.com/pocketbase/dbx"
	"github.com/pocketbase/pocketbase/core"
)

func visibilityTestState(t *testing.T, size int) (core.App, string, string, []string) {
	t.Helper()
	app := newSchemaTestApp(t)
	if err := ensureSchema(app); err != nil {
		t.Fatal(err)
	}
	user, token := createRouteTestUser(t, app, "visibility")
	ids := make([]string, size)
	if err := app.RunInTransaction(func(txApp core.App) error {
		for i := range size {
			ids[i] = fmt.Sprintf("visibility%05d", i)
			record := newSubscriptionPerformanceCoreRecord(t, txApp, user.Id, subscriptionPerformanceRecord{
				ID: ids[i], Name: ids[i], Price: "12", Currency: "USD", BillingCycle: "monthly", Category: "productivity", Status: "active",
				StartDate: "2026-01-01", NextBillingDate: "2027-01-01", ReminderDays: -1, Tags: []string{"Team"},
			})
			if err := txApp.Save(record); err != nil {
				return err
			}
		}
		return rebuildSubscriptionDerivedStateForUser(txApp, user.Id, time.Now())
	}); err != nil {
		t.Fatal(err)
	}
	registerRecordHooks(app)
	return app, user.Id, token, ids
}

func visibilityBody(t *testing.T, ids []string, target, preview bool) string {
	t.Helper()
	body, err := json.Marshal(map[string]interface{}{"selection": map[string]interface{}{"ids": ids}, "publicHidden": target, "dryRun": preview})
	if err != nil {
		t.Fatal(err)
	}
	return string(body)
}

func TestPublicVisibilityBulkSizes(t *testing.T) {
	for _, size := range []int{5, 100, 500, 5000} {
		t.Run(fmt.Sprint(size), func(t *testing.T) {
			app, userID, token, ids := visibilityTestState(t, size)
			before := readSubscriptionDerivedOracleSnapshot(t, app, userID)
			for _, target := range []bool{true, false} {
				response := serveTestRequest(t, app, http.MethodPost, "/api/app/subscriptions/bulk-public-visibility", visibilityBody(t, ids, target, false), token)
				if response.Code != 200 {
					t.Fatalf("HTTP %d: %s", response.Code, response.Body.String())
				}
				var envelope struct {
					Data subscriptionBulkPublicVisibilityResponse `json:"data"`
				}
				if err := json.Unmarshal(response.Body.Bytes(), &envelope); err != nil {
					t.Fatal(err)
				}
				if envelope.Data.ChangedCount != size || envelope.Data.MatchedCount != size || len(envelope.Data.FailedIDs) != 0 {
					t.Fatalf("wrong result: %+v", envelope.Data)
				}
				var count struct {
					Count int `db:"count"`
				}
				if err := app.DB().NewQuery(`SELECT COUNT(*) AS count FROM subscriptions s JOIN subscription_list_index i ON i.subscription_id=s.id
					WHERE s.user={:user} AND s.publicHidden={:target} AND i.public_hidden=s.publicHidden AND s.updated=i.updated_at`).Bind(dbx.Params{"user": userID, "target": boolToSQLiteInt(target)}).One(&count); err != nil {
					t.Fatal(err)
				}
				if count.Count != size {
					t.Fatalf("facts/index mismatch: %d", count.Count)
				}
			}
			after := readSubscriptionDerivedOracleSnapshot(t, app, userID)
			if !reflect.DeepEqual(before, after) {
				t.Fatal("unrelated derived data changed")
			}
			ops, err := measureSubscriptionDBOperations(app, func() error {
				rows, err := readPublicVisibilityTargets(app, userID, subscriptionBulkPublicVisibilitySelection{IDs: ids}, "2026-09-15")
				if err != nil {
					return err
				}
				return writePublicVisibility(app, userID, rows, true)
			})
			if err != nil {
				t.Fatal(err)
			}
			if ops.WriteStatements != 2 || ops.DerivedWrites != 1 {
				t.Fatalf("unbounded writes: %+v", ops)
			}
			t.Logf("rows=%d reads=%d writes=%d elapsed=%s allocated=%d", size, ops.ReadQueries, ops.WriteStatements, ops.Elapsed, ops.AllocatedBytes)
			if size == 500 {
				for _, statement := range ops.ReadSQL {
					var plan []struct {
						Detail string `db:"detail"`
					}
					if err := app.DB().NewQuery("EXPLAIN QUERY PLAN " + statement).All(&plan); err != nil {
						t.Fatal(err)
					}
					details := []string{}
					for _, row := range plan {
						details = append(details, row.Detail)
					}
					joined := strings.Join(details, "\n")
					if !strings.Contains(joined, "SEARCH s USING INDEX") || !strings.Contains(joined, "id=?)") || strings.Contains(joined, "SCAN s ") || strings.HasSuffix(joined, "SCAN s") {
						t.Fatalf("visibility must point-read IDs: %s", joined)
					}
				}
			}
		})
	}
}

func TestPublicVisibilityPreviewIdempotencyAndOwner(t *testing.T) {
	app, userID, token, ids := visibilityTestState(t, 5)
	foreign, _ := createRouteTestUser(t, app, "visibility-foreign")
	other := createRouteTestSubscription(t, app, foreign.Id, nil)
	selection := append([]string{ids[0], ids[0]}, other.Id, "missing")
	for _, preview := range []bool{true, false, false} {
		r := serveTestRequest(t, app, http.MethodPost, "/api/app/subscriptions/bulk-public-visibility", visibilityBody(t, selection, true, preview), token)
		if r.Code != 200 {
			t.Fatalf("HTTP %d: %s", r.Code, r.Body.String())
		}
		var response struct {
			Data subscriptionBulkPublicVisibilityResponse `json:"data"`
		}
		if err := json.Unmarshal(r.Body.Bytes(), &response); err != nil {
			t.Fatal(err)
		}
		if response.Data.MatchedCount != 1 || !reflect.DeepEqual(response.Data.FailedIDs, []string{other.Id, "missing"}) {
			t.Fatalf("invalid owner result: %+v", response.Data)
		}
		rows, err := readPublicVisibilityTargets(app, userID, subscriptionBulkPublicVisibilitySelection{IDs: ids}, "2026-09-15")
		if err != nil {
			t.Fatal(err)
		}
		if preview && rows[0].PublicHidden {
			t.Fatal("dry run mutated facts")
		}
	}
	current, err := app.FindRecordById("subscriptions", other.Id)
	if err != nil || current.GetBool("publicHidden") {
		t.Fatal("foreign subscription modified")
	}
	for _, body := range []string{`{"selection":{"ids":null,"categories":["expired"]},"publicHidden":true}`, `{"selection":{"ids":["x"],"categories":null},"publicHidden":true}`, `{"selection":{"ids":null},"publicHidden":true}`, `{"selection":{"ids":[]},"publicHidden":true}`, `{"selection":{"ids":[""]},"publicHidden":true}`, `{"selection":{"ids":["x"]}}`, `{"selection":{"ids":["x"]},"publicHidden":null}`, `{"selection":{"ids":["x"]},"publicHidden":true,"dryRun":null}`, `{"selection":{"ids":["x"],"categories":[]},"publicHidden":true}`, `{"selection":{"ids":["x"],"unknown":true},"publicHidden":true}`} {
		r := serveTestRequest(t, app, http.MethodPost, "/api/app/subscriptions/bulk-public-visibility", body, token)
		if r.Code != 400 {
			t.Fatalf("expected strict 400: %s: %d %s", body, r.Code, r.Body.String())
		}
	}
}

func TestPublicVisibilityTransactionRollbackAndConflict(t *testing.T) {
	app, userID, token, ids := visibilityTestState(t, 500)
	rows, err := readPublicVisibilityTargets(app, userID, subscriptionBulkPublicVisibilitySelection{IDs: ids}, "2026-09-15")
	if err != nil {
		t.Fatal(err)
	}
	if _, err := app.DB().NewQuery(`CREATE TRIGGER fail_visibility BEFORE UPDATE ON subscription_list_index WHEN OLD.subscription_id='visibility00499' BEGIN SELECT RAISE(ABORT, 'injected failure'); END`).Execute(); err != nil {
		t.Fatal(err)
	}
	response := serveTestRequest(t, app, http.MethodPost, "/api/app/subscriptions/bulk-public-visibility", visibilityBody(t, ids, true, false), token)
	if response.Code != 500 {
		t.Fatalf("expected 500, got %d %s", response.Code, response.Body.String())
	}
	current, err := readPublicVisibilityTargets(app, userID, subscriptionBulkPublicVisibilitySelection{IDs: ids}, "2026-09-15")
	if err != nil || !reflect.DeepEqual(rows, current) {
		t.Fatal("failed batch did not roll back facts")
	}
	if _, err := app.DB().NewQuery("DROP TRIGGER fail_visibility; UPDATE subscriptions SET status='paused' WHERE id='visibility00000'").Execute(); err != nil {
		t.Fatal(err)
	}
	if err := writePublicVisibility(app, userID, rows, true); !errors.Is(err, errSubscriptionWriteConflict) {
		t.Fatalf("expected snapshot conflict, got %v", err)
	}
	response = serveTestRequest(t, app, http.MethodPost, "/api/app/subscriptions/bulk-public-visibility", visibilityBody(t, ids, true, false), token)
	if response.Code != 200 {
		t.Fatalf("retry failed: %d %s", response.Code, response.Body.String())
	}
	single := serveTestRequest(t, app, http.MethodPatch, "/api/app/subscriptions/"+ids[0], `{"publicHidden":false}`, token)
	if single.Code != 200 {
		t.Fatalf("single PATCH failed: %d %s", single.Code, single.Body.String())
	}
}

File: apps/worker/src/subscription-public-visibility.test.ts
import { beforeEach, describe, expect, it, vi } from "vitest";
import { bulkUpdatePublicVisibility } from "./subscription-public-visibility";
import { rebuildSubscriptionDerivedStateForUser } from "./subscription-derived-state";
import { openDerivedStateDatabase, subscriptionRow, insertSubscriptionStatement, USER_ID, NOW } from "./subscription-d1-test-support";
import { subscriptionBulkPublicVisibilityResponseSchema } from "@renewlet/shared/schemas/subscriptions";

vi.mock("./auth", () => ({ requireAuth: vi.fn(async () => ({ user: { id: USER_ID } })) }));
beforeEach(() => vi.restoreAllMocks());

function request(selection: unknown, publicHidden = true, dryRun = false) {
  return new Request("https://renewlet.test/api/app/subscriptions/bulk-public-visibility", {
    method: "POST", headers: { "content-type": "application/json" }, body: JSON.stringify({ selection, publicHidden, dryRun }),
  });
}

async function seed(size: number) {
  const state = openDerivedStateDatabase();
  const ids = Array.from({ length: size }, (_, i) => `sub_visibility_${i}`);
  for (const id of ids) await insertSubscriptionStatement(state.env, subscriptionRow(id)).run();
  await rebuildSubscriptionDerivedStateForUser(state.env, USER_ID, NOW);
  return { ...state, ids };
}

function snapshot(state: ReturnType<typeof openDerivedStateDatabase>) {
  return Object.fromEntries(["subscriptions", "subscription_list_index", "subscription_tags", "subscription_repeat_schedule", "subscription_user_stats", "subscription_scheduler_state"]
    .map((table) => [table, state.db.prepare(`SELECT * FROM ${table} ORDER BY 1`).all()]));
}

describe("bulk public visibility with real SQLite transactions", () => {
  it.each([5, 100, 500, 5000])("hides and restores %i rows with bounded reads and a single atomic batch", async (size) => {
    const state = await seed(size);
    try {
      const before = snapshot(state);
      const batch = vi.spyOn(state.env.DB, "batch");
      const prepare = vi.spyOn(state.env.DB, "prepare");
      for (const target of [true, false]) {
        batch.mockClear(); prepare.mockClear();
        const response = await bulkUpdatePublicVisibility(request({ ids: state.ids }, target), state.env);
        expect(response.status).toBe(200);
        expect(subscriptionBulkPublicVisibilityResponseSchema.parse(await response.json()).data).toEqual({ matchedCount: size, changedCount: size, skippedCount: 0, failedIds: [] });
        expect(batch).toHaveBeenCalledTimes(1);
        expect(batch.mock.calls[0]?.[0]).toHaveLength(4);
        expect(prepare.mock.calls.filter(([sql]) => sql.includes("SELECT s.id, s.user_id, s.public_hidden"))).toHaveLength(Math.ceil(size / 400));
        expect(state.db.prepare("SELECT COUNT(*) AS count FROM subscriptions WHERE public_hidden = ?").get(Number(target))?.["count"]).toBe(size);
        expect(state.db.prepare("SELECT COUNT(*) AS count FROM subscriptions s JOIN subscription_list_index i ON i.subscription_id=s.id WHERE s.public_hidden != i.public_hidden OR s.updated_at != i.updated_at").get()?.["count"]).toBe(0);
      }
      const after = snapshot(state);
      for (const table of ["subscription_tags", "subscription_repeat_schedule", "subscription_user_stats", "subscription_scheduler_state"]) expect(after[table]).toEqual(before[table]);
      batch.mockClear();
      const unchanged = await bulkUpdatePublicVisibility(request({ ids: state.ids }, false), state.env);
      expect(subscriptionBulkPublicVisibilityResponseSchema.parse(await unchanged.json()).data).toMatchObject({ changedCount: 0, skippedCount: size });
      expect(batch).not.toHaveBeenCalled();
    } finally { state.db.close(); }
  });

  it("deduplicates IDs, keeps their order and reports missing/foreign IDs without updating them", async () => {
    const state = await seed(3);
    try {
      state.db.exec("INSERT INTO users (id) VALUES ('foreign')");
      await insertSubscriptionStatement(state.env, subscriptionRow("foreign-sub", { user_id: "foreign" })).run();
      const before = snapshot(state);
      const response = await bulkUpdatePublicVisibility(request({ ids: [state.ids[2], state.ids[0], state.ids[0], "missing", "foreign-sub"] }, true, true), state.env);
      expect(subscriptionBulkPublicVisibilityResponseSchema.parse(await response.json()).data).toEqual({ matchedCount: 2, changedCount: 2, skippedCount: 0, failedIds: ["missing", "foreign-sub"] });
      expect(snapshot(state)).toEqual(before);
      const apply = await bulkUpdatePublicVisibility(request({ ids: [state.ids[0], "foreign-sub"] }), state.env);
      expect(subscriptionBulkPublicVisibilityResponseSchema.parse(await apply.json()).data.failedIds).toEqual(["foreign-sub"]);
      expect(state.db.prepare("SELECT public_hidden FROM subscriptions WHERE id='foreign-sub'").get()?.["public_hidden"]).toBe(0);
    } finally { state.db.close(); }
  });

  it.each([0, 199, 499])("rolls back all facts and indexes when row %i fails", async (failureIndex) => {
    const state = await seed(500);
    const log = vi.spyOn(console, "error").mockImplementation(() => undefined);
    try {
      const before = snapshot(state);
      state.db.exec(`CREATE TRIGGER fail_visibility BEFORE UPDATE ON subscription_list_index WHEN OLD.subscription_id = 'sub_visibility_${failureIndex}' BEGIN SELECT RAISE(ABORT, 'injected index failure'); END`);
      await expect(bulkUpdatePublicVisibility(request({ ids: state.ids }), state.env)).rejects.toMatchObject({ status: 500, code: "SUBSCRIPTION_PUBLIC_VISIBILITY_FAILED" });
      expect(snapshot(state)).toEqual(before);
      expect(log).toHaveBeenCalledWith("subscription_public_visibility_failed", expect.objectContaining({ requestId: expect.any(String) }));
    } finally { state.db.close(); }
  });

  it("detects missing indexes before write and ignored index updates before commit", async () => {
    const state = await seed(5);
    vi.spyOn(console, "error").mockImplementation(() => undefined);
    try {
      state.db.exec("CREATE TRIGGER ignore_visibility BEFORE UPDATE ON subscription_list_index BEGIN SELECT RAISE(IGNORE); END");
      const before = snapshot(state);
      await expect(bulkUpdatePublicVisibility(request({ ids: state.ids }), state.env)).rejects.toMatchObject({ status: 500 });
      expect(snapshot(state)).toEqual(before);
      state.db.exec("DROP TRIGGER ignore_visibility; DELETE FROM subscription_list_index WHERE subscription_id='sub_visibility_0'");
      const missing = snapshot(state);
      await expect(bulkUpdatePublicVisibility(request({ ids: state.ids }), state.env)).rejects.toMatchObject({ status: 500 });
      expect(snapshot(state)).toEqual(missing);
    } finally { state.db.close(); }
  });

  it("returns 409 for a changed snapshot and retries using the current facts", async () => {
    const state = await seed(5);
    try {
      const execute = state.env.DB.batch.bind(state.env.DB);
      vi.spyOn(state.env.DB, "batch").mockImplementationOnce((statements) => {
        state.db.exec("UPDATE subscriptions SET status='paused', updated_at='2026-09-15T12:00:00Z' WHERE id='sub_visibility_0'");
        return execute(statements);
      });
      await expect(bulkUpdatePublicVisibility(request({ ids: state.ids }), state.env)).rejects.toMatchObject({ status: 409, code: "SUBSCRIPTION_WRITE_CONFLICT" });
      expect(state.db.prepare("SELECT SUM(public_hidden) AS hidden FROM subscriptions").get()?.["hidden"]).toBe(0);
      expect((await bulkUpdatePublicVisibility(request({ ids: state.ids }), state.env)).status).toBe(200);
    } finally { state.db.close(); }
  });

  it("uses ID lookups for target reads and snapshot guards even when an owner index exists", async () => {
    const state = await seed(100);
    try {
      const prepare = vi.spyOn(state.env.DB, "prepare");
      await bulkUpdatePublicVisibility(request({ ids: state.ids }), state.env);
      const guardSQL = prepare.mock.calls.find(([sql]) => sql.includes("FROM (WITH expected"))?.[0];
      expect(guardSQL).toBeDefined();
      if (!guardSQL) throw new Error("Missing snapshot guard SQL");
      const plan = state.db.prepare(`EXPLAIN QUERY PLAN ${guardSQL}`).all(1, 1, "[]", USER_ID);
      expect(plan.some((row) => String(row["detail"]).includes("SEARCH s USING INDEX sqlite_autoindex_subscriptions_1 (id=?)"))).toBe(true);
      const readSQL = prepare.mock.calls.find(([sql]) => sql.includes("SELECT s.id, s.user_id, s.public_hidden"))?.[0];
      if (!readSQL) throw new Error("Missing target read SQL");
      const readPlan = state.db.prepare(`EXPLAIN QUERY PLAN ${readSQL}`).all("[]", USER_ID);
      expect(readPlan.some((row) => String(row["detail"]).includes("SEARCH s USING INDEX sqlite_autoindex_subscriptions_1 (id=?)"))).toBe(true);
    } finally { state.db.close(); }
  });

  it("evaluates categories at apply time and excludes paused/cancelled and fixed-term buyouts", async () => {
    const state = await seed(6);
    try {
      state.db.exec(`UPDATE subscriptions SET next_billing_date='2000-01-01';
        UPDATE subscriptions SET status='paused' WHERE id='sub_visibility_1';
        UPDATE subscriptions SET status='cancelled' WHERE id='sub_visibility_2';
        UPDATE subscriptions SET billing_cycle='one-time', one_time_term_count=0, next_billing_date='2099-01-01' WHERE id='sub_visibility_3';
        UPDATE subscriptions SET billing_cycle='one-time', one_time_term_count=12, one_time_term_unit='month', next_billing_date='2099-01-01' WHERE id='sub_visibility_4';
        UPDATE subscriptions SET next_billing_date='2099-01-01' WHERE id='sub_visibility_5'`);
      const preview = await bulkUpdatePublicVisibility(request({ categories: ["expired", "lifetime"] }, true, true), state.env);
      expect(subscriptionBulkPublicVisibilityResponseSchema.parse(await preview.json()).data.matchedCount).toBe(2);
      state.db.exec("UPDATE subscriptions SET status='paused' WHERE id='sub_visibility_0'");
      const apply = await bulkUpdatePublicVisibility(request({ categories: ["expired", "lifetime"] }), state.env);
      expect(subscriptionBulkPublicVisibilityResponseSchema.parse(await apply.json()).data.changedCount).toBe(1);
      expect(state.db.prepare("SELECT id FROM subscriptions WHERE public_hidden=1").all()).toEqual([{ id: "sub_visibility_3" }]);
    } finally { state.db.close(); }
  });

  it.each([{ ids: [] }, { ids: [""] }, { ids: ["id"], categories: ["expired"] }, { ids: ["id"], unknown: true }, { ids: Array.from({ length: 5001 }, () => "id") }])("rejects invalid selections before reads: %j", async (selection) => {
    const state = openDerivedStateDatabase();
    try {
      const prepare = vi.spyOn(state.env.DB, "prepare");
      await expect(bulkUpdatePublicVisibility(request(selection), state.env)).rejects.toMatchObject({ status: 400 });
      expect(prepare).not.toHaveBeenCalled();
    } finally { state.db.close(); }
  });
});
